x/observer/keeper/ballot.go

```diff
@@ -58,10 +71,25 @@ func (k Keeper) GetAllBallots(ctx sdk.Context) (voters []*types.Ballot) {
 
 // AddBallotToList adds a ballot to the list of ballots for a given height.
 func (k Keeper) AddBallotToList(ctx sdk.Context, ballot types.Ballot) {
-	list, found := k.GetBallotList(ctx, ballot.BallotCreationHeight)
+	list, found := k.GetBallotListForHeight(ctx, ballot.BallotCreationHeight)
 	if !found {
 		list = types.BallotListForHeight{Height: ballot.BallotCreationHeight, BallotsIndexList: []string{}}
 	}
 	list.BallotsIndexList = append(list.BallotsIndexList, ballot.BallotIdentifier)
 	k.SetBallotList(ctx, &list)
 }
+
+// ClearMaturedBallotsAndBallotList deletes all matured ballots and the list of ballots for a given height.
+// It also emits an event for each ballot deleted.
+func (k Keeper) ClearMaturedBallotsAndBallotList(ctx sdk.Context, ballots []types.Ballot, maturityBlocks int64) {
```

Comment:
I agree with @swift1337 , we should add more encapsulation, the external module shouldn't need to provide a list of ballot

-> You provide an height
-> You can get all ballots from that height in the function
-> You delete all the ballots

Reply:
I don't think you need to reread the store for ballots? The ballot list from height give the ID list that can be directly used in the delete function?

This is small optimization in any case I would consider the encapsulation of the code more important

Reply:
That is true, but the delete ballot event uses other fields of the ballot. 

We could alternatively fire the event from abci.go, but I feel that firing it from the for loop is better

Reply:
The current interface doesn't respect encapsulation principle. We delete ballots from a list and a `BallotListForHeight` value, while the list of ballots can be derived from this height. Both values should be deleted atomically from the same inputs, the current interface allows an external module to create an inconsistent state for the ballot